8303904: [macos]Transparent Windows Paint Wrong (Opaque, Pixelated) w/o Volatile Buffering

mickleness opened the pull request:
When "swing.volatileImageBufferEnabled" is false: we were mistakenly using image at 100% resolution.

~~In hindsight the original ticket probably should be split up into two distinct issues:~~
~~1. The window is opaque, so pixels that should be transparent are black.~~
~~2. The window is the wrong resolution. On a 200% resolution monitor it renders as if it were 100% (so it looks pixelated).~~

Previously this PR also addressed opacity, but that has been resolved separately via JDK-8382201. Now this PR focuses exclusively on the resolution / AffineTransform of the image we paint to. This PR refactors some existing code (the BackingStoreMultiResolutionImage) from JViewport to wrap a BufferedImage in a smaller (transformed) MultiResolutionImage.

This PR started 2 years ago. I got stuck and abandoned it, and @anass-baya picked it back up again this year: https://github.com/openjdk/jdk/pull/23430/

In that PR @mrserb suggested we try to fix this problem in RepaintManager (see https://github.com/openjdk/jdk/pull/23430#discussion_r2089555453 ), so that's what this PR does.

---------
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Error
&nbsp;⚠️ Pull request body is missing required line: `- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).`

### Issue
 * [JDK-8303904](https://bugs.openjdk.org/browse/JDK-8303904): [macos]Transparent Windows Paint Wrong (Opaque, Pixelated) w/o Volatile Buffering (**Bug** - P3)


### Contributors
 * Anass Baya `<abaya@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13196/head:pull/13196` \
`$ git checkout pull/13196`

Update a local copy of the PR: \
`$ git checkout pull/13196` \
`$ git pull https://git.openjdk.org/jdk.git pull/13196/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13196`

View PR using the GUI difftool: \
`$ git pr show -t 13196`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13196.diff">https://git.openjdk.org/jdk/pull/13196.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13196#issuecomment-1485907307)
</details>

mickleness committed:
Merge pull request #1 from openjdk/master

Merge openjdk/jdk into mickleness/jdk

mickleness committed:
Merge pull request #2 from openjdk/master

Merge openjdk/jdk

mickleness committed:
Merge pull request #3 from openjdk/master

Merge from openjdk/jdk

mickleness committed:
Merge pull request #4 from openjdk/master

Updating mickleness/jdk from openjdk/jdk

mickleness committed:
8303904: Add failing unit test

mickleness committed:
8303904: use Transparency.TRANSLUCENT for non-opaque windows

mickleness committed:
8303904: change createAcceleratedImage to return a potentially high-res image

This is part of the original write-up, but it might deserve to be split up into its own ticket.

mickleness committed:
8303904: rolling back prev createAcceleratedImage commit

I predict once I start a code review about 8308904 that if we want to address the resolution of createAcceleratedImage: that can be a separate ticket. This change feels much more invasive to me; I wouldn't be surprised if some existing code out there is already casting the return value of createAcceleratedImage(..) as a BufferedImage.

mickleness committed:
8303904: updating instruction text to reflect narrowed focus of this branch

mickleness committed:
8303904: adding failing unit test for OpenGL on Mac

I started to notice some intermittent failures in the Robot test, so now I'm having it wait a few beats on the event dispatch thread before grabbing the pixel color.

Also I'm adding a more abstract check-the-translucency-of-the-BufferedImage output. If this `testImageTransparency` passes and `testPixelColor` fails: that's a red flag that maybe the screen isn't ready when the Robot grabs the pixel.

mickleness committed:
8303904: resolve failing bug8308904_OGL test on Mac

This is the same solution previously applied to the MTLGraphicsConfig

mickleness committed:
8303904: apply same solution to other GraphicsConfigurations

I don't have my setup configured to test these, but I confirmed they all appear to support translucency in their getColorModel(int) method.

mickleness committed:
8303904: cleaning up comment

mickleness committed:
8303904: fixing compiler `cannot find symbol` errors

mickleness commented:
Alan, I may be confused. Does VAqua rely on turning off volatile buffering?

Or your comment might (?) make more sense if it's for my other PR ([for 8303950](https://github.com/openjdk/jdk/pull/12993)) that relates to this method:

https://github.com/openjdk/jdk/blob/a8871f5d26e5cb42c031c7b736ec30b1b147a2bc/src/java.desktop/share/classes/java/awt/Window.java#L3946-L3960

(Also: thanks for mentioning VAqua! I'll check it out. I was a big fan of Werner's Quaqua work a decade ago at another job.)

mickleness committed:
8303904: avoid System.exit(1)

mrserb recommended against this in a separate PR

https://github.com/openjdk/jdk/pull/13408#discussion_r1162182212

mickleness commented:
What is the fate of auto-closed PRs?

That is: as long as I don't delete this branch in my repo will this be available for future reference if anyone dusts off this ticket in the future?

aivanov-jdk commented:
> What is the fate of auto-closed PRs?

@mickleness You can reopen the PR with the [`/open`](https://wiki.openjdk.org/display/SKARA/Pull+Request+Commands#PullRequestCommands-/open) command.

> That is: as long as I don't delete this branch in my repo will this be available for future reference if anyone dusts off this ticket in the future?

If you keep the branch in your repo, you or anyone could use it as the base to continue working on this issue.

mickleness committed:
Merge branch 'master' into JDK-8303904

mickleness committed:
8303904: move MultiResolutionImage logic to RepaintManager

This is in response to:
https://github.com/openjdk/jdk/pull/23430/files#r2089555453

mickleness committed:
8303904: support rotated GraphicsConfiguration

In my setup: monitors only vary in resolution (scaleX, scaleY). But I've seen some old OpenJDK tickets that (I think?) allude to GraphicsConfigurations with AffineTransforms that include rotation.

This should support rotation as long as it comes in increments of 90 degrees.

mickleness committed:
8303904: fixing AffineTransform fields

mickleness committed:
8303904: rewrite tests as automated

Also now they iterate over more configurations: they'll test every available monitor with an opaque and translucent window. I tested using my laptop monitor (200%), an external monitor (200%), and another external monitor (100%)

mickleness committed:
8303904: code cleanup

Removing old changes that don't apply, & updating copyright year.

mickleness commented:
/open

mickleness committed:
8303904: refactor BackingStoreMultiResolutionImage

Convert this private a private internal class to its own package private class

mickleness committed:
Revert "8303904: fixing AffineTransform fields"

This reverts commit 2d32aaf0e670a9967877e5a4bf4c0ec1079bb070.

mickleness committed:
Revert "8303904: support rotated GraphicsConfiguration"

This reverts commit 55b1a708cdc8608824d04974ad96b62fd50149d3.

mickleness committed:
8303904: use BackingStoreMultiResolutionImage

It's functionally the same thing; this is just reusing code (and it avoids the hacky fake image involving getScaledInstance).

mickleness committed:
8303904: minor cleanup

This makes this method ALWAYS return a BackingStoreMultiResolutionImage, even if the scaled size is the same as the virtual size. (Just because making this method return a BufferedImage OR a BackingStoreMultiResolutionImage seems like an unnecessary level of abstraction; and that could turn into a potential point of confusion for future devs.)

mickleness committed:
8303904: minor cleanup

mickleness committed:
8303904: minor cleanup

mickleness committed:
8303904: designing around imagined edge cases

I don't know if these will ever come up in the real world, but since this is critical infrastructure I want to be careful. (A Swing app may become unusable if this code fails.)

Concern A:
In the event our AffineTransform is a rotation or a flip:
Now we'll use the c.createImage(virtualWidth, virtualHeight), which means we fall back to the code we've been using for over a decade.

Concern B:
In the event our AffineTransform scales to zero, we'll at least make the image 1x1.

mickleness committed:
8303904: comment cleanup

mickleness commented:
/contributor @anass-baya

mickleness commented:
/contributor add abaya

mickleness committed:
Merge branch 'master' into JDK-8303904

mickleness committed:
Merge branch 'master' into JDK-8303904

mickleness committed:
8303904: Updating copyright year

mickleness committed:
Merge branch 'master' into JDK-8303904

# Conflicts:
#	src/java.desktop/macosx/classes/sun/java2d/metal/MTLGraphicsConfig.java
#	src/java.desktop/macosx/classes/sun/java2d/opengl/CGLGraphicsConfig.java
#	src/java.desktop/share/classes/javax/swing/JViewport.java

mickleness commented:
/template append
